### ISSUE TITLE
PhraseMatcher still only using the last label when multiple components are defined

### ISSUE THREAD

**Issue by ChrisPalmerNZ** [explosion/spaCy#4610]
I may not correctly understand how to add multiple Matcher components, but after previously reporting this and having observed it agreed as being a bug and subsequently marked as resolved, I have waited until a newer version of spaCy has arrived (2.2.1) and have re-tested. Again, I am having a situation where if I have more than one phrase matcher in the pipeline only the label of the last matcher is used for any entities matched.

Please let me know if I am making a mistake in adding the multiple components, or if this is a bug and has been fixed then in which version of spaCy, and how I can update to it. 
 
## How to reproduce the behaviour

```
import spacy
print("spaCy version:", spacy.__version__)
# 2.2.1

from spacy.matcher import PhraseMatcher
from spacy.tokens import Doc, Span

nlp = spacy.load('en_core_web_sm')

# Method to handle the custom login names component
def login_name_component(doc):
    # Apply the matcher to the doc
    matches = matcher(doc)
    # Create a Span for each match and assign the label 'LOGIN_NAME'
    spans = [Span(doc, start, end, label="LOGIN_NAME") for match_id, start, end in matches]
    # Overwrite the doc.ents with the matched spans
    doc.ents = spans
    return doc

# Method to handle the custom State names component
def state_terr_component(doc):
    # Apply the matcher to the doc
    matches = matcher(doc)
    # Create a Span for each match and assign the label 'STATE_TERR'
    spans = [Span(doc, start, end, label="STATE_TERR") for match_id, start, end in matches]
    # Overwrite the doc.ents with the matched spans
    doc.ents = spans
    return doc

# Method to print document entities
def print_ents(txt):
    test_doc = nlp(txt)
    for ent in test_doc.ents:
        # Print the entity text and its label
        print(ent.text, ent.label_)

txt = "JimBrown - 10/11/2019 - Note: Christopher Robin visited ACT then NSW on 08 Nov 2019\r\nSonnyBill - 11/11/2019 - Noted"

print()
print("Text being processed:")
print(txt)
print()
print("Entities before any phrase matcher components:")
print_ents(txt)

# Add login names phrase matcher
login_names = ['JimBrown', 'SonnyBill']
login_name_pattern = list(nlp.pipe(login_names))
login_name_pattern
matcher = PhraseMatcher(nlp.vocab)
matcher.add('LOGIN_NAME', None, *login_name_pattern )

nlp.add_pipe(login_name_component, before="ner")
print()
print("login_name_pattern added :", login_name_pattern)
print("Pipe names after adding login_name_component")
print(nlp.pipe_names)

# login names correctly identified
print()
print("Entities after addition of login_name_component:")
print_ents(txt)

# Add State / Territory phrase matcher
State_Terr  = ['ACT', 'NSW', 'NT', 'QLD', 'SA', 'TAS', 'VIC', 'WA']
state_terr_pattern = list(nlp.pipe(State_Terr))

matcher.add('STATE_TERR', None, *state_terr_pattern)
# Add the component to the pipeline before the 'login_name_component'
nlp.add_pipe(state_terr_component, before="login_name_component")

print()
print("state_terr_pattern added :", state_terr_pattern)
print("Pipe names after adding state_terr_component")
print(nlp.pipe_names)

# State Territory names incorrectly identified as login names
print()
print("Entities after addition of state_terr_component:")
print_ents(txt)
```
## Your Environment

* Operating System: Windows-7-6.1.7601-SP1
* Python Version Used: 3.6.9
* spaCy Version Used: 2.2.1

**Comment by adrianeboyd** [comment]
Hi, I don't think this is a bug. The main problems:

* your components refer to a global `matcher` that includes both patterns at the end of the script
* your components hard-code the label on the span so the second component matches both patterns but you manually label all matches as LOGIN
* each component is overwriting any ents earlier components have added

You're going to have difficulties if components refer to global variables that are changing state and you (probably) do not want each component to clobber the previous component's labels.

The typical way to do this would be:

```
matcher = PhraseMatcher(nlp.vocab)
matcher.add('LOGIN_NAME', None, *login_name_pattern )
matcher.add('STATE_TERR', None, *state_terr_pattern)
matches = matcher(doc)
spans = [Span(doc, start, end, label=match_id) for match_id, start, end in matches]
# note that this overwrites any existing entities added by other components
doc.ents = spans
```

**Comment by ChrisPalmerNZ** [comment]
Thanks for this. I will retry following your advice. 

I am confused though about where it is that I "manually label all matches as LOGIN". I have labelled the login_name_pattern as LOGIN_NAME and the state_terr_pattern as STATE_TERR.  

Also, can you please explain why it is wrong to have a global matcher that has more than one pattern - I thought that the idea of the matcher is to be able to build up successive patterns. And, why is your suggested code _not_ global whereas mine is?

Finally, is there a complete example anywhere of how to do what I am attempting, I had to cobble this together based on code fragments I found on the spaCy site, and obviously have not understood  it well...

**Comment by adrianeboyd** [comment]
Given that you want a pipeline component that turns matches into entities, I would suggest using the `EntityRuler`, which is intended for exactly this kind of task:

https://spacy.io/usage/rule-based-matching#entityruler

**Comment by ChrisPalmerNZ** [comment]
Thanks, I will give that a go...

**Comment by ChrisPalmerNZ** [comment]
Hi @adrianeboyd, thanks for your helpful suggestions! 

I have reviewed all of this, and yes the Entity Ruler approach works for me. And - even better for the purposes of answering my inquiry, I understood why you've stated that my call to `login_name_component` acted as a global replacement of entity labels with LOGIN_NAME. 

Instead of having ` spans = [Span(doc, start, end, label="LOGIN_NAME") for match_id, start, end in matches]` ... 

I just needed to implement your suggestion of `spans = [Span(doc, start, end, label=match_id) for match_id, start, end in matches]` in both of my handlers to get the desired result!

```
JimBrown LOGIN_NAME
Christopher Robin PERSON
ACT STATE_TERR
NSW STATE_TERR
08 Nov 2019 CARDINAL
SonnyBill LOGIN_NAME
11/11/2019 DATE
```
